### PR TITLE
Added listener to the MapReduceRuntimeService to record the state change...

### PR DESCRIPTION
**Pull request by sagarkapare**
...s.

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java [hunk 1]
```diff
@@ -156,53 +156,48 @@ public ProgramController run(final Program program, ProgramOptions options) {
     final Service mapReduceRuntimeService = new MapReduceRuntimeService(cConf, hConf, mapReduce, spec, context,
                                                                         program.getJarLocation(), locationFactory,
                                                                         streamAdmin, txSystemClient);
-    final ProgramController controller = new MapReduceProgramController(mapReduceRuntimeService, context);
-    controller.addListener(new AbstractListener() {
+    mapReduceRuntimeService.addListener(new Service.Listener() {
       @Override
-      public void init(ProgramController.State state, @Nullable Throwable cause) {
+      public void starting() {
         store.setStart(program.getId(), runId.getId(), TimeUnit.SECONDS.convert(System.currentTimeMillis(),
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java]: Extract the timestamp from the RunId. `RunIds.getTimeMillis(runId)`.

Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java]: Also, you can add method to `RunIds` to simplify the conversion. E.g. `RunIds.getTime(RunId runId, TimeUnit unit)`.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java [hunk 2]
```diff
@@ -156,53 +156,48 @@ public ProgramController run(final Program program, ProgramOptions options) {
...
-          error(controller.getFailureCause());
-        }
       }
 
       @Override
-      public void completed() {
-        store.setStop(program.getId(), runId.getId(),
-                      TimeUnit.SECONDS.convert(System.currentTimeMillis(), TimeUnit.MILLISECONDS),
-                      ProgramController.State.COMPLETED.getRunStatus());
+      public void running() {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java]: You can extend from `ServiceListenerAdapter` to simplify implementation for no-op method.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java [hunk 3]
```diff
@@ -156,53 +156,48 @@ public ProgramController run(final Program program, ProgramOptions options) {
...
       @Override
-      public void suspended() {
-        store.setSuspend(program.getId(), runId.getId());
-      }
-
-      @Override
-      public void resuming() {
-        store.setResume(program.getId(), runId.getId());
+      public void terminated(Service.State from) {
+        if (from != Service.State.STOPPING) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java]: Usually it's easier to understand if 'positive' check is used.

```java
if (from == STOPPING) {
} else {
}
```

**Comment by chtyim** [comment]
Just couple comments.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java [hunk 4]
```diff
@@ -156,53 +156,48 @@ public ProgramController run(final Program program, ProgramOptions options) {
...
+                        ProgramController.State.KILLED.getRunStatus());
+        }
       }
 
       @Override
-      public void error(Throwable cause) {
-        LOG.info("MapReduce Program stopped with error {}, {}", program.getId(), runId, cause);
+      public void failed(Service.State from, Throwable failure) {
         store.setStop(program.getId(), runId.getId(),
                       TimeUnit.SECONDS.convert(System.currentTimeMillis(), TimeUnit.MILLISECONDS),
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java]: I find it simpler and easier to read if you use `TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis())`

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java (hunk 1 above)
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java]: Added `RunIds.getTime(RunId runId, TimeUnit unit)` mthod to the RunIds class.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java (hunk 2 above)
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java]: Using `ServiceListenerAdapter`

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java (hunk 3 above)
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java]: Fixed.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java (hunk 4 above)
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java]: Changed to used `TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis())`

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/runtime/RunIds.java [hunk 5]
```diff
@@ -77,12 +78,13 @@ public static RunId generate(long timeMillis) {
   }
 
   /**
-   * @return time in milliseconds from the UUID if it is a time-based UUID, -1 otherwise.
+   * @return time from the UUID if it is a time-based UUID, -1 otherwise.
    */
-  public static long getTimeMillis(RunId runId) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/runtime/RunIds.java]: Why remove this method? you shouldn't. I was suggesting to add the `getTime`, not to replace this one.

Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/runtime/RunIds.java]: ohh ok. Will add it. But not sure who will call `getTimeMillis`, as all calls to this method are for setting the store and I thought they will use the `getTime`.   
`AppMetadataStore` was also calling `getTimeMillis` but it was converting the returned value to the seconds.

**Comment by sagarkapare** [comment]
Build passed here: http://builds.cask.co/browse/CDAP-DUT1366-2

**Comment by chtyim** [comment]
LGTM

**Pull request merged**